File: server/cpp/write_named_pipe.cpp
#include "write_named_pipe.h"
#include "named_pipe.h"
#include "constants.h"
#include "exceptions.h"
#include <sys/stat.h>  // for mkfifo
#include <fcntl.h>  // for open
#include <string>
using namespace std;

WriteNamedPipe::WriteNamedPipe(bool create)
{
	int status;

	if (create)
	{
		status = mkfifo(FIFO_NAME, S_IFIFO | 0666);
		if (status == -1)
			throw runtime_error("NamedPipe::NamedPipe::mkfifo");
	}

	m_fifofd = open(FIFO_NAME, O_WRONLY);
	if (m_fifofd == -1)
		throw runtime_error("NamedPipe::NamedPipe::open");
}

void WriteNamedPipe::writePipe(const string& text, unsigned len) const
{
	int status;

	status = write(m_fifofd, text.c_str(), len);
	if (status == -1)
		throw runtime_error("NamedPipe::write::write");
}

File: python-server/write_named_pipe.py
from named_pipe import NamedPipe
import os
import fcntl

class WriteNamedPipe(NamedPipe):

    def __init__(self, fifo_name, create = True):
        if create:
            try:
                os.mkfifo(fifo_name)
            except:
                raise RuntimeError("WriteNamedPipe::WriteNamedPipe::mkfifo")
        try:
            self.fifofd = open(fifo_name, "w+", 0)
            fd = self.fifofd.fileno()
            flag = fcntl.fcntl(fd, fcntl.F_GETFL)
            fcntl.fcntl(fd, fcntl.F_SETFL, flag | os.O_NONBLOCK)
        except:
            raise RuntimeError("WriteNamedPipe::WriteNamedPipe::open")

    def write_pipe(self, text):
        try:
            self.fifofd.write(text)
        except:
            raise RuntimeError("WriteNamedPipe::write::write")
